### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by michalc** [hughsk/from2#12]
Calling `next` seems to re-call the read function immediately, adding to the call stack. This limits the logic you can put into the read function,

For example, the below program ends up throwing `RangeError: Maximum call stack size exceeded`, rather than adding a single character to the stream and then exiting.

    var from = require('from2');
    var stream = from(function(size, next) {
      next(null, '-');

      // Never gets here
      process.exit(0);
    });
    stream.read();

I'm certainly not an expert on Node streams, but this seems unexpected to me.

**Comment by yoshuawuyts** [comment]
hah, looks like you've unleashed Zalgo. Fwiw I think this behavior is expected when doing synchronous stuff using streams -> `next()` gives no guarantees about asynchronicity. To prevent this behavior every `next()` call could be wrapped in a `process.nextTick()` / `dezalgo` call, but that would hurt perf. You're free to wrap it yourself though.

tl;dr: I believe it's expected when doing sync stuff.

**Comment by michalc** [comment]
I guess if it's expected, then fair enough.

In case anyone stumbles on this in future, the below code, using raw stream.Readable, works in the way I expected the original code to work.

    var source = stream.Readable();
    source._read = function() {
      this.push('-');
      this;push(null);
      console.log('Hello');
    };
    source.read();

**Comment by yoshuawuyts** [comment]
@michalc yeah, but the second example is different from the first. The code you're sharing does the following:
- push the `-` character
- terminate the stream
- log `'Hello'`

Which is different from pushing an infinite number sequence synchronously.

**Comment by michalc** [comment]
Completely agreed and understood about the second code being different from the first. I posted it exactly because *is* different!

**Comment by yoshuawuyts** [comment]
Good ol' Zalgo. Now that you've modified your first code example I see what you're doing; this is slightly annoying, perhaps this package _should_ dezalgo by default.

This works tho:
```js
var from = require('from2');
var done = false
var source = from(function(size, next) {
  if (done) return next(null, null);
  done = true
  next(null, '-');


  console.log('Hello');
});
source.pipe(process.stdout)
```
This will fail
```js
var from = require('from2');
var done = false
var source = from(function(size, next) {
  if (done) return next(null, null);
  next(null, '-');
  done = true

  console.log('Hello');
});
source.pipe(process.stdout)
```
There's points to be made for both sides; though I agree it's confusing.

**Comment by mafintosh** [comment]
what about renaming `next` to `callback` in the docs? that kinda implies that you should only call it once

**Comment by yoshuawuyts** [comment]
yeah, I reckon that's better. I think it'll still be confusing tho; looking back at the code it takes me a bit of crunching before seeing what's going on - if this was production code I wouldn't have caught any bugs by simply :eyes: it
